### PR TITLE
Refactoring common ajax request code into storage.

### PR DESCRIPTION
**Pull request by moniika**

### CODE REVIEW

File: appengine/common/storage.js [hunk 1]
```diff
@@ -60,71 +60,112 @@ BlocklyStorage.restoreBlocks = function() {
...
+ * @param {?Function=} opt_onFailure Function to call after request completes
+ *    unsuccessfully. Defaults to BlocklyStorage alert of request status.
+ * @param {string=} [opt_method='POST'] The HTTP request method to use.
+ */
+BlocklyStorage.makeFormRequest =
+    function(form, opt_onSuccess, opt_onFailure, opt_method) {
+  var data = [];
+  for (var i = 0, element; (element = form.elements[i]); i++) {
+    if (element.name) {
+      data[i] = encodeURIComponent(element.name) + '=' +
```
Comment (NeilFraser) [inline appengine/common/storage.js]: This should be .push() since we might skip an element.

File: appengine/common/storage.js [hunk 2]
```diff
@@ -60,71 +60,112 @@ BlocklyStorage.restoreBlocks = function() {
...
-  BlocklyStorage.xhr_ = new XMLHttpRequest();
-  BlocklyStorage.xhr_.name = name;
-  BlocklyStorage.xhr_.onload = BlocklyStorage.handleRequest_;
-  BlocklyStorage.xhr_.open('POST', url);
-  BlocklyStorage.xhr_.setRequestHeader('Content-Type',
-      'application/x-www-form-urlencoded');
-  BlocklyStorage.xhr_.send(name + '=' + encodeURIComponent(content));
+  BlocklyStorage.xhr_[url] = new XMLHttpRequest();
+  BlocklyStorage.xhr_[url].onload = function() {
+    if (this.status === 200) {
```
Comment (NeilFraser) [inline appengine/common/storage.js]: == is shorter.

File: appengine/common/storage.js [hunk 3]
```diff
@@ -60,71 +60,112 @@ BlocklyStorage.restoreBlocks = function() {
...
 
 /**
- * Global reference to current AJAX request.
- * @type XMLHttpRequest
- * @private
+ * Global reference to current AJAX requests.
+ * @type Object.<string, XMLHttpRequest>
  */
-BlocklyStorage.xhr_ = null;
+BlocklyStorage.xhr_ = {};
```
[the remark this answers is not in the document]
Comment (moniika) [reply appengine/common/storage.js]: Done.

File: appengine/common/storage.js [hunk 4]
```diff
@@ -60,71 +60,112 @@ BlocklyStorage.restoreBlocks = function() {
...
- * @private
+ * @param {Array.<string>} data Body of data to be sent in request.
+ * @param {?Function=} opt_onSuccess Function to call after request completes
+ *    successfully.
+ * @param {?Function=} opt_onFailure Function to call after request completes
+ *    unsuccessfully. Defaults to BlocklyStorage alert of request status.
+ * @param {string=} [opt_method='POST'] The HTTP request method to use.
+ */
+BlocklyStorage.makeFormRequest =
+    function(form, opt_onSuccess, opt_onFailure, opt_method) {
```
[the remark this answers is not in the document]
Comment (moniika) [reply appengine/common/storage.js]: Updated.

File: appengine/common/storage.js [hunk 5]
```diff
@@ -60,71 +60,112 @@ BlocklyStorage.restoreBlocks = function() {
...
-  BlocklyStorage.xhr_.onload = BlocklyStorage.handleRequest_;
-  BlocklyStorage.xhr_.open('POST', url);
-  BlocklyStorage.xhr_.setRequestHeader('Content-Type',
-      'application/x-www-form-urlencoded');
-  BlocklyStorage.xhr_.send(name + '=' + encodeURIComponent(content));
+  BlocklyStorage.xhr_[url] = new XMLHttpRequest();
+  BlocklyStorage.xhr_[url].onload = function() {
+    if (this.status === 200) {
+      if (opt_onSuccess) {
+        opt_onSuccess.call(this);
```
[the remark this answers is not in the document]
Comment (moniika) [reply appengine/common/storage.js]: Done.

File: appengine/common/storage.js (hunk 1 above)
Comment (moniika) [reply appengine/common/storage.js]: Good point.

File: appengine/common/storage.js (hunk 2 above)
Comment (moniika) [reply appengine/common/storage.js]: I was putting "===" to appease my IDE warnings about type cohesion. It's technically in the style guide, but its the "new" one (that also applies to ES6).
If strong preference for "==" I can change.

**Comment by NeilFraser** [comment]
Thanks for the updates.  My only question is regarding the usage of `BlocklyStorage.makeFormRequest`.  Currently it is only used in lib-gallery.

A quick (but incomplete) browse through the 'pond' branch doesn't show any other usages of this function.  If it will only be used by the gallery, then it should be moved to lib-gallery.

**Comment by moniika** [comment]
> Thanks for the updates. My only question is regarding the usage of `BlocklyStorage.makeFormRequest`. Currently it is only used in lib-gallery.
> 
> A quick (but incomplete) browse through the 'pond' branch doesn't show any other usages of this function. If it will only be used by the gallery, then it should be moved to lib-gallery.

We would be using it in the pond-multi branch, though I'm not sure whether the need would be permanent. We make use of forms for request but its currently for testing purposes. 
I lean towards leaving it where it is now and moving it later if it turns out its definitely not needed.

**Comment by NeilFraser** [comment]
> We would be using it in the pond-multi branch, though I'm not sure whether the need would be permanent. We make use of forms for request but its currently for testing purposes.
> I lean towards leaving it where it is now and moving it later if it turns out its definitely not needed.

My concern is that by placing it in storage.js, this function is being sent for every single game, even though most of them don't use it.  Furthermore, storage.js is not compressed.  And storage.js is a blocking load at the beginning.  This adds 1KB to every page.

The less we have in storage.js, the better.

**Comment by moniika** [comment]
> > We would be using it in the pond-multi branch, though I'm not sure whether the need would be permanent. We make use of forms for request but its currently for testing purposes.
> > I lean towards leaving it where it is now and moving it later if it turns out its definitely not needed.
> 
> My concern is that by placing it in storage.js, this function is being sent for every single game, even though most of them don't use it. Furthermore, storage.js is not compressed. And storage.js is a blocking load at the beginning. This adds 1KB to every page.
> 
> The less we have in storage.js, the better.

Makes sense. I've moved it to lib-gallery.

**Pull request merged**